### ISSUE TITLE
java.lang.IllegalArgumentException: Truncate length should be positive

### ISSUE THREAD

**Issue by changquanyou** [apache/iceberg#2065]
Env: 
- spark3.0 
- Iceberg0.10.0
- `./bin/spark-sql --packages org.apache.iceberg:iceberg-spark3-runtime:0.10.0 \
--conf spark.sql.catalog.hadoop_prod=org.apache.iceberg.spark.SparkCatalog \
--conf spark.sql.catalog.hadoop_prod.type=hadoop \
--conf spark.sql.catalog.hadoop_prod.warehouse=hdfs://bfd-wfj-67p121:8020/iceberg/warehouse \
--master spark://bfd-wfj-67p123:7077`
Issue: Using Spark Sql:  select title  from mf_data where pubDayStr = '2021-01-09' and **title like  'apple%'** limit 1;
Here is the Exception:
21/01/11 18:02:43 ERROR SparkSQLDriver: Failed in [select title  from mf_data where pubDayStr = '2021-01-09' and title like  'apple%' limit 1]
java.lang.IllegalArgumentException: Truncate length should be positive
	at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
	at org.apache.iceberg.util.BinaryUtil.truncateBinary(BinaryUtil.java:35)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator$MetricsEvalVisitor.startsWith(InclusiveMetricsEvaluator.java:326)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator$MetricsEvalVisitor.startsWith(InclusiveMetricsEvaluator.java:76)
	at org.apache.iceberg.expressions.ExpressionVisitors$BoundExpressionVisitor.predicate(ExpressionVisitors.java:135)
	at org.apache.iceberg.expressions.ExpressionVisitors.visitEvaluator(ExpressionVisitors.java:322)
	at org.apache.iceberg.expressions.ExpressionVisitors.visitEvaluator(ExpressionVisitors.java:341)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator$MetricsEvalVisitor.eval(InclusiveMetricsEvaluator.java:99)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator$MetricsEvalVisitor.access$100(InclusiveMetricsEvaluator.java:76)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator.eval(InclusiveMetricsEvaluator.java:70)
	at org.apache.iceberg.ManifestReader.lambda$entries$0(ManifestReader.java:182)
	at org.apache.iceberg.io.CloseableIterable$3.shouldKeep(CloseableIterable.java:82)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:67)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:50)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:65)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:50)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:65)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:50)
	at org.apache.iceberg.io.CloseableIterable$4$1.hasNext(CloseableIterable.java:108)
	at org.apache.iceberg.util.ParallelIterable$ParallelIterator.lambda$new$1(ParallelIterable.java:67)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IllegalArgumentException: Truncate length should be positive
	at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
	at org.apache.iceberg.util.BinaryUtil.truncateBinary(BinaryUtil.java:35)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator$MetricsEvalVisitor.startsWith(InclusiveMetricsEvaluator.java:326)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator$MetricsEvalVisitor.startsWith(InclusiveMetricsEvaluator.java:76)
	at org.apache.iceberg.expressions.ExpressionVisitors$BoundExpressionVisitor.predicate(ExpressionVisitors.java:135)
	at org.apache.iceberg.expressions.ExpressionVisitors.visitEvaluator(ExpressionVisitors.java:322)
	at org.apache.iceberg.expressions.ExpressionVisitors.visitEvaluator(ExpressionVisitors.java:341)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator$MetricsEvalVisitor.eval(InclusiveMetricsEvaluator.java:99)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator$MetricsEvalVisitor.access$100(InclusiveMetricsEvaluator.java:76)
	at org.apache.iceberg.expressions.InclusiveMetricsEvaluator.eval(InclusiveMetricsEvaluator.java:70)
	at org.apache.iceberg.ManifestReader.lambda$entries$0(ManifestReader.java:182)
	at org.apache.iceberg.io.CloseableIterable$3.shouldKeep(CloseableIterable.java:82)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:67)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:50)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:65)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:50)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:65)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:50)
	at org.apache.iceberg.io.CloseableIterable$4$1.hasNext(CloseableIterable.java:108)
	at org.apache.iceberg.util.ParallelIterable$ParallelIterator.lambda$new$1(ParallelIterable.java:67)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

**Comment by kbendick** [comment]
You know I was actually just working on this PR https://github.com/apache/iceberg/pull/2062 where I encountered this and was debating whether or not I should fix it when it came up (but I didn't want to pack too much into an admittedly large PR already).

I have a fix stashed away that I can push if you'd like. 👍

**Comment by kbendick** [comment]
Though if you'd like to submit a patch, by all means let me know 🙂 . Otherwise, I can take this.

**Comment by kbendick** [comment]
@changquanyou Is it possible that you could share the individual row where this occurred, as well as the table's partition spec?

As I mentioned, I have a patch for it, but I'd like to be sure that I understand the situation under which it occurred.

For example, I imagine we want to be sure that we aren't allowing people to create partition specs for columns with `truncate[0]` (which I'd have to check to see if that's possible). But it seems like an individual row with an empty string is what caused this... I'm guessing either pubDayStr or title, where maybe non-parseable values / values of the incorrect type in in the raw data are possibly being stored in an empty string partition.

**Comment by changquanyou** [comment]
@kbendick  thanks for your reply, I will explain in detail：

- Table Schema:
  `CREATE TABLE mf_data (
    urlHash STRING,
    titleSimHash STRING,
    docId STRING,
    title STRING,
    content STRING,
    pubTime LONG,
    pubDayStr STRING
    ) using iceberg **PARTITIONED BY** (pubDayStr);` 
   The Service assurance that  partition key  is not empty. the title may be empty. 
- Data Source：
  Consume from Kafka Topic, And sink to IceBerg Table. At once, the data is normal。  
- Experiments:
 1. select a record randomly
     `select title from mf_data where pubDayStr = '2021-01-09' limit 1;` 
       Here is the  response: 喜讯 | 花都这个基地被认定为省级高新区
 2. use like query on the  **title** field
     `select title from mf_data where pubDayStr = '2021-01-09' and title like '喜讯%'  limit 1;`
      **java.lang.IllegalArgumentException: Truncate length should be positive** occurred
 4. use like query on the **content** field:
     `select title from mf_data where pubDayStr = '2021-01-09' and content like '喜讯%'  limit 1;`
       The result is normal, None Exception occurred
 5.  use Spark SQL client insert:
      `insert into mf_data values('7a1ab0138b535572ff346801c8b61ec0','778ec08c0e9930a64fa3ea03de1334aa','bfd_c68722c6-2224-41f8-82bd-675f0b81f0cd','房东整租霍营小区二层两居室','房东整租霍营小区二层两居室',1607912994000,'2020-12-14');`
`select title from mf_data where pubDayStr = '2020-12-14' and content like '房东%'  limit 1;`
The above result is normal, of course  I Insert another data（the title is empty） :
  `insert into mf_data values('testb0138b535572ff346801c8b61ec0','test08c0e9930a64fa3ea03de1334aa','bfd_c68722c6-2224-41f8-82bd-675f0b81test1','','Test房东整租霍营小区二层两居室',1607912994001,'2020-12-14');` 
 Use The Same Spark Sql Query, **Truncate length should be positive**  is occurred.
- TODO:
   Due to the huge table, It‘s not easy to confirm the empty title exists. then I will append the title length flag to the table.

From the Experiment Result, if the field size is zero, It does happen  IllegalArgumentException. I hope that will help you.

**Comment by kbendick** [comment]
Oh thank you so much. This is wonderful information. And I completely understand that the table could be too large to determine any single row that would cause this issue.

I've discussed with @RussellSpitzer on this and I'm not going to include this in my current open PR that I mentioned where I first encountered this issue, as that PR is already large enough and I don't want to block it from getting merged. But I will take a look. I believe that I have a relatively simple fix for this and I will use these characters you've provided as test cases if you don't mind. I don't speak any languages that don't use the normal latin alphabet, with the exception of characters found in Romance languages like like ñ, œ, ü, ç, etc so it's helpful to have sample data like this. To me, it's important to have more test cases with much more extended code points from utf-8 and non-ascii letters.

However, I think that your issue can be solved relatively easily and likely does not relate to the non-ascii characters. But I'll take a look at the non-ascii characters to verify as well, while prioritizing your issue first given that it's affecting what appears to be production queries.

I'll work on this and should have something ready in the next few days. Luckily, I'm off this week so I've got time. Thanks again for the report!
